apps/gateway/src/routes/api/v1/document-logs/:documentLogUuid/handlers/addMessage.ts

```diff
@@ -0,0 +1,42 @@
+import { zValidator } from '@hono/zod-validator'
+import { addMessages, LogSources } from '@latitude-data/core'
+import { messageSchema } from '$/common/messageSchema'
+import { pipeToStream } from '$/common/pipeToStream'
+import { queues } from '$/jobs'
+import { Factory } from 'hono/factory'
+import { streamSSE } from 'hono/streaming'
+import { z } from 'zod'
+
+const factory = new Factory()
+
+const schema = z.object({
+  messages: z.array(messageSchema),
+  source: z.nativeEnum(LogSources).optional().default(LogSources.API),
+})
+
+export const addMessageHandler = factory.createHandlers(
+  zValidator('json', schema),
+  async (c) => {
+    return streamSSE(c, async (stream) => {
+      const { documentLogUuid } = c.req.param()
+      const { messages, source } = c.req.valid('json')
+      const workspace = c.get('workspace')
+      const apiKey = c.get('apiKey')
+
+      const result = await addMessages({
+        workspace,
+        documentLogUuid,
+        messages,
+        providerLogHandler: (log) => {
+          queues.defaultQueue.jobs.enqueueCreateProviderLogJob({
+            ...log,
+            source,
+            apiKeyId: apiKey.id,
+          })
+        },
+      }).then((r) => r.unwrap())
+
+      await pipeToStream(stream, result.stream)
+    })
+  },
+)
```

Comment:
If we can add new messages to a chat, I think it may be interesting to mark which of the `providerLogs` is the last one of the chan, and which are from an afterwards added message.

An example of this is "time until response", which is the time spent until getting the stream of the latest step in a chain. To estimate this, I am currently doing `documentLogDuration` - `lastProviderLogDuration`, but this would no longer work when new messages are added afterwards.

Reply:
Captured here https://github.com/latitude-dev/latitude-llm/issues/120